File: programming_beyond_the_basics/introduction/luhn_algorithm.clj
(defn lookup [digit]
  (if (> digit  9) (- digit 9) digit))

(defn map-digit [index-digit-vec]
  (let [index (nth index-digit-vec  0)
        digit (Character/digit (nth index-digit-vec 1) 10)]
    (if (odd? index) (lookup (* digit 2)) digit)))

(defn verify [digits]
  (let [sum (reduce  + (doall (map map-digit (map-indexed vector (reverse digits)))))]
    (if (= (mod sum 10) 0) true false)))

(println (verify "17893729975"))
(println (verify "17893729974"))

File: programming_beyond_the_basics/introduction/python/luhn_algorithm.py
def verify(digits):
    double = lambda num: num * 2 
    sum_digits = lambda num: sum(map(int, str(num)))
    double_and_sum = lambda num: sum_digits(double(num))
    conditional_double_and_sum = lambda tuple : (lambda index, digit: double_and_sum(digit) if index%2 else digit)(tuple[0], int(tuple[1]))
    return sum(map(conditional_double_and_sum, enumerate(digits))) % 10 == 0


if __name__ == '__main__'
    import os
    os.system('clear')
    print('------------')
    print(verify("17893729975"))
    print(verify("17893729974"))
    print('ok')
    print('------------')

